openday_scavenger/api/puzzles/service.py

```diff
@@ -296,11 +309,103 @@ def record_access(db_session: Session, puzzle_name: str, visitor_uid: str) -> Ac
+) -> dict[str, Any]:
+    """
+    Get the state information for the given puzzle and visitor.
+
+    Args:
+        db_session (Session): The SQLAlchemy session object.
+        puzzle_name (str): The name of the puzzle the visitor accessed.
+        visitor_auth (VisitorAuth): The authenticated visitor that accessed the puzzle.
+
+    Returns:
+        dict[str, Any]: The state as a dictionary that the puzzle can structure as it sees fit.
+    """
+
+    # Get the database model for the visitor.
+    visitor = db_session.query(Visitor).filter(Visitor.uid == visitor_auth.uid).first()
+    if visitor is None:
+        raise VisitorUIDInvalidError(f"Could not find visitor {visitor_auth.uid} in the database.")
+
+    # Get the database models for the puzzle.
+    puzzle = get(db_session, puzzle_name)
+
+    # Check whether a state for the puzzle and visitor already exists.
+    state_model = (
+        db_session.query(State)
+        .filter(State.puzzle_id == puzzle.id, State.visitor_id == visitor.id)
+        .first()
+    )
+
+    if state_model is None:
+        return {}
+    else:
+        return json.loads(state_model.state)
+
+
+def set_puzzle_state(
+    db_session: Session, *, puzzle_name: str, visitor_auth: VisitorAuth, state: dict[str, Any]
+):
+    """
+    Set the state information for the given puzzle and visitor.
+
+    Args:
+        db_session (Session): The SQLAlchemy session object.
+        puzzle_name (str): The name of the puzzle the visitor accessed.
+        visitor_auth (VisitorAuth): The authenticated visitor that accessed the puzzle.
+        state (dict[str, Any]): The state as a dictionary that the puzzle can structure as it sees fit.
+    """
+
+    # Get the database model for the visitor.
+    visitor = db_session.query(Visitor).filter(Visitor.uid == visitor_auth.uid).first()
+    if visitor is None:
+        raise VisitorUIDInvalidError(f"Could not find visitor {visitor_auth.uid} in the database.")
+
+    # Get the database models for the puzzle.
+    puzzle = get(db_session, puzzle_name)
+
+    # Check whether a state for the puzzle and visitor already exists.
+    state_model = (
+        db_session.query(State)
+        .filter(State.puzzle_id == puzzle.id, State.visitor_id == visitor.id)
+        .first()
+    )
+
+    # If the state doesn't exist yet, create it, otherwise overwrite the state information
+    if state_model is None:
+        state_model = State(
+            puzzle=puzzle, visitor=visitor, updated_at=datetime.now(), state=json.dumps(state)
+        )
+        try:
+            db_session.add(state_model)
+            db_session.commit()
+        except Exception:
+            db_session.rollback()
+            raise PuzzleStateCreationError(
+                f"Could not create a state for {puzzle.name} by {visitor.uid}"
+            )
+    else:
+        state_model.updated_at = datetime.now()
+        state_model.state = json.dumps(state)
```

Comment:
What would you think of being extra-careful and putting a jsonable_encoder in front of json.dumps?
`state_model.state = json.dumps(jsonable_encoder(state))`
In case something in state is not straight-up jsonable